src/useAddonState/server.ts

```diff
@@ -0,0 +1,88 @@
+import type { Channel } from "@storybook/channels";
+
+import { getValue, GetValuePayload, setValue, SetValuePayload } from "./common";
+
+class AddonState {
+  private values: Record<string, any> = {};
+
+  private listeners: Record<string, ((value: any) => unknown)[]> = {};
+
+  get(key: string) {
+    return this.values[key];
+  }
+
+  set(key: string, value: any) {
+    this.values[key] = value;
+    (this.listeners[key] || []).forEach((l) => l(value));
+  }
+
+  listen<T>(key: string, cb: (value: T) => unknown) {
+    this.listeners[key] = [...(this.listeners[key] || []), cb];
+  }
+
+  // for testing
+  reset() {
+    this.values = {};
+    this.listeners = {};
+  }
+}
+
+const addonState = new AddonState();
+
+let listening = false;
+function ensureListening(channel: Pick<Channel, "emit" | "on">) {
+  if (listening) return;
+
+  channel.on(getValue, ({ key }: GetValuePayload) => {
+    const value = addonState.get(key);
+    if (value !== undefined) channel.emit(setValue, { key, value } satisfies SetValuePayload);
+  });
+
+  channel.on(setValue, ({ key, value }: SetValuePayload) => addonState.set(key, value));
+
+  listening = true;
+}
+
+const requestedKeys = new Set<string>();
+// For testing
+export function resetAddonState() {
+  addonState.reset();
+  requestedKeys.clear();
+  listening = false;
+}
+
+function apiGetAddonState<T>(channel: Pick<Channel, "emit" | "on">, key: string) {
+  // Ask the universe for the value of the key if we've never seen it before
+  if (!requestedKeys.has(key)) {
+    channel.emit(getValue, { key } satisfies GetValuePayload);
+    requestedKeys.add(key);
+  }
+
+  return addonState.get(key) as T | undefined;
+}
+
+function apiSetAddonState<T>(channel: Pick<Channel, "emit" | "on">, key: string, value: T) {
+  addonState.set(key, value);
+
+  channel.emit(setValue, { key, value } as SetValuePayload);
+}
+
+export function useAddonState<T>(channel: Pick<Channel, "emit" | "on">, key: string) {
+  ensureListening(channel);
+
+  return {
+    get value() {
+      return apiGetAddonState(channel, key);
+    },
+
+    set value(newValue: T) {
+      apiSetAddonState(channel, key, newValue);
+    },
```

Comment:
Why the getter/setter? I'm not digging the `addonState.value = ...` syntax personally (it's just preference though).

Reply:
Very happy to change the API, I'm to wedded to it at all.